pilot/pkg/serviceregistry/aggregate/controller.go

```diff
@@ -228,11 +228,25 @@ func (c *Controller) GetProxyServiceInstances(node *model.Proxy) ([]*model.Servi
 	// It doesn't make sense for a single proxy to be found in more than one registry.
 	// TODO: if otherwise, warning or else what to do about it.
 	for _, r := range c.GetRegistries() {
+		// Skip the service registry when we know there won't be a match
+		// because the proxy is in a different cluster. We can't trust
+		// the default service registry because its always named `Kubernetes`
+		// and won't match the CLUSTER_ID label from the proxies.
+		if r.Cluster() != "" &&
+			r.Cluster() != string(serviceregistry.Kubernetes) &&
+			node.Metadata != nil &&
+			node.Metadata.ClusterID != "" &&
+			node.Metadata.ClusterID != r.Cluster() {
+			log.Debugf("GetProxyServiceInstances(): not checking cluster %v: proxy %v is associated with cluster %v",
+				r.Cluster(), node.ID, node.Metadata.ClusterID)
+			continue
+		}
+
 		instances, err := r.GetProxyServiceInstances(node)
 		if err != nil {
 			errs = multierror.Append(errs, err)
 		} else if len(instances) > 0 {
-			out = append(out, instances...)
+			out = instances
```

Comment:
Why this change? This means we can only get endpoints from one registry? Which will break all the other registries like ServiceEntry?

Reply:
See the for loop comment - "It doesn't make sense for a single proxy to be found in more than one registry".  `out` only ever has at most instances from one registry. The code breaks out of the loop two lines below on line 251.

Reply:
I don't think that comment is accurate? I am surprised all tests pass, I think you should be able to have ServiceEntry + Kube.

@rshriram ?

Reply:
I've removed the change from this PR. We can address it in a follow-up.